src/ser.rs

```diff
@@ -691,7 +691,7 @@ impl<'a> Serializer<'a> {
 
         match *state {
             State::Table { first, .. } => {
-                if !first.get() {
+                if !first.get() || !self.dst.is_empty() {
```

Comment:
Just a few questions, I'm not terribly familiar with the serializer side of the library (I'm having a bit of trouble understanding what `first` actually means here).

Should `State::Array` also have a similar check?

Is it even necessary to check if `first` is set?  When would `dst` be empty and `first` be false?

Reply:
I tried to make as little changes as possible so I kept the `first` check. I can always try out without it if you want?

> Should `State::Array` also have a similar check?

I don't think we want it for arrays (at least I think so?). The point of this change is to handle this case:

```toml
[foo]
x = 1

[foo.something]
y = 2
```

In the current code, there wouldn't have the empty line between the two whereas it should. I think that an array shouldn't have the empty line, no?

Reply:
The reason I ask about arrays is because the following:

```rust
assert_eq!(
    Table(map! {
        "test" => Table(map!{"val" => Integer(2)}),
         "z" => Table(map!{
            "foo" => Array(vec![Table(map!{"hello" => Integer(3)})])
         })
    })
    .to_string(),
    "[test]\n\
    val = 2\n\
    [[z.foo]]\n\
    hello = 3\n"
);
```

Notice there is no empty line between the `[test]` table and the `[[z.foo]]` array.  This also seems to depend on the alphabetic ordering (`[[a.foo]]` does include a blank line).

We can wait to see if Alex has any opinion on how it should work, since it does seem a little wonky.

Reply:
Oh dear it's been years at this point since I read this code. I agree with @ehuss though that arrays/tables need to be treated the same here, and I don't think checking `self.dst` is right because you could also be serializing into a non-empty string to start with or something like that, but we should produce the same serialization regardless